src/boot/efi/stub.c

```diff
@@ -357,14 +356,69 @@ static void install_addon_devicetrees(
         }
 }
 
+static inline void iovec_array_append(struct iovec **arr, size_t *n_arr, struct iovec elem) {
+        *arr = xrealloc(*arr, *n_arr * sizeof(struct iovec), (*n_arr + 1)  * sizeof(struct iovec));
+        (*arr)[(*n_arr)++] = elem;
+}
+
+static void measure_ucode_addons_and_create_final_list(
+                struct iovec initrds[static _INITRD_MAX],
+                struct iovec **all_initrds,
+                size_t *n_all_initrds,
+                NamedAddon *ucode_addons,
+                size_t n_ucode_addons,
+                int *sections_measured) {
+        EFI_STATUS err;
+
+        assert(initrds);
+        assert(all_initrds);
+        assert(n_all_initrds);
+        assert(ucode_addons);
+        assert(n_ucode_addons);
+        assert(sections_measured);
+
+        /* Populate all_initrds by first adding ucode addons, and then copying all the static initrds */
+
+        /* Ucode addons need to be measured and copied into all_initrds in reverse order,
+         * the kernel takes the first one it finds. */
+        for (ssize_t i = n_ucode_addons - 1; i >= 0; i--) {
+                bool m = false;
+                err = tpm_log_tagged_event(
+                                TPM2_PCR_KERNEL_CONFIG,
+                                POINTER_TO_PHYSICAL_ADDRESS(ucode_addons[i].blob.iov_base),
+                                ucode_addons[i].blob.iov_len,
+                                UCODE_ADDON_EVENT_TAG_ID,
+                                ucode_addons[i].filename,
+                                &m);
+                if (err != EFI_SUCCESS)
+                        return (void) log_error_status(
+                                        err,
+                                        "Unable to extend PCR %i with UCODE addon '%ls': %m",
+                                        TPM2_PCR_KERNEL_CONFIG,
+                                        ucode_addons[i].filename);
+
+                combine_measured_flag(sections_measured, m);
+
+                iovec_array_append(all_initrds, n_all_initrds, ucode_addons[i].blob);
+        }
+
+        /* Now append the other initrds from the static list */
+        FOREACH_ARRAY(i, initrds, _INITRD_MAX) {
+                if (iovec_is_set(i))
```

Comment:
maybe move this check into iovec_array_append(), so that you do an early exit there if the iovec is no set.